plugins/policy-recommendation/policy_recommendation_job.py

```diff
@@ -0,0 +1,839 @@
+                    _from = [antrea_crd.NetworkPolicyPeer(
+                        pod_selector = kubernetes.client.V1LabelSelector())]
+                )],
+            )
+        )
+        policies.append(dict_to_yaml(acnp.to_dict()))
+    return policies
+
+def generate_sql_query(table_name, limit, start_time, end_time, unprotected):
+    sql_query = "SELECT {} FROM {}".format(", ".join(FLOW_TABLE_COLUMNS), table_name)
+    if unprotected:
+        sql_query += " WHERE ingressNetworkPolicyName == '' AND egressNetworkPolicyName == ''"
+    else:
+        # Select user trusted denied flows when unprotected equals False
+        sql_query += " WHERE trusted == 1"
+    if start_time:
+        sql_query += " AND flowEndSeconds >= '{}'".format(start_time)
+    if end_time:
+        sql_query += " AND flowEndSeconds < '{}'".format(end_time)
+    sql_query += " GROUP BY {}".format(", ".join(FLOW_TABLE_COLUMNS))
+    if limit:
+        sql_query += " LIMIT {}".format(limit)
+    return sql_query
+
+def read_flow_df(spark, db_jdbc_address, sql_query, rm_labels):
+    flow_df = spark.read \
+        .format("jdbc") \
+        .option("driver", "ru.yandex.clickhouse.ClickHouseDriver") \
+        .option("url", db_jdbc_address) \
+        .option("user", os.getenv("CH_USERNAME")) \
+        .option("password", os.getenv("CH_PASSWORD")) \
+        .option("query",  sql_query)\
+        .load()
+    if rm_labels:
+        flow_df = flow_df.withColumn('sourcePodLabels', udf(remove_meaningless_labels, StringType())("sourcePodLabels"))\
+            .withColumn('destinationPodLabels', udf(remove_meaningless_labels, StringType())("destinationPodLabels"))\
+            .dropDuplicates(['sourcePodLabels', 'destinationPodLabels'])
+    flow_df = flow_df.withColumn('flowType', udf(get_flow_type, StringType())("destinationServicePortName", "destinationPodLabels"))
+    return flow_df
+
+def write_recommendation_result(spark, result, recommendation_type, db_jdbc_address, table_name, id):
+    if not id:
+        recommendation_id = str(uuid.uuid4())
+    else:
+        recommendation_id = id
+    result_dict = {
+        'id': recommendation_id,
+        'type': recommendation_type,
+        'timeCreated': datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
+        'yamls': '---\n'.join(filter(None, result))
+    }
+    result_df = spark.createDataFrame([result_dict])
+    result_df.write\
+        .mode("append") \
+        .format("jdbc") \
+        .option("driver", "ru.yandex.clickhouse.ClickHouseDriver") \
+        .option("url", db_jdbc_address) \
+        .option("user", os.getenv("CH_USERNAME")) \
+        .option("password", os.getenv("CH_PASSWORD")) \
+        .option("dbtable", table_name) \
+        .save()
+    return recommendation_id
+
+def initial_recommendation_job(spark, db_jdbc_address, table_name, limit=0, option=1, start_time=None, end_time=None, ns_allow_list=NAMESPACE_ALLOW_LIST, rm_labels=False, to_services=True):
```

Comment:
At some point we need apply pep and pylint to this - there are a lot of violation - but can also be done in a subsequent iteration.

Reply:
Agree, pep/pylint could be done in a Github workflow. I write it down as a TODO too.

Reply:
thansk!